src/backend/joanie/core/api/client.py

```diff
@@ -65,68 +69,65 @@ class CourseProductRelationViewSet(
-        ).select_related("course", "product")
-
-
-class ProductViewSet(mixins.RetrieveModelMixin, viewsets.GenericViewSet):
-    """API ViewSet for all interactions with products."""
-
-    lookup_field = "id"
-    permissions_classes = [drf_permissions.AllowAny]
-    filterset_class = filters.ProductViewSetFilter
-    queryset = models.Product.objects.all()
-    serializer_class = serializers.ProductSerializer
-
-    def filter_queryset(self, queryset):
-        """
-        Custom queryset to limit to products actually related to the course given in querystring.
-        """
-        queryset = super().filter_queryset(queryset)
+    def get_serializer_context(self):
+        """Pass authenticated username into the serializer context."""
+        return {
+            "username": self.username,
+        }
 
+    def get_permissions(self):
+        """Anonymous user should be able to retrieve a course product relation."""
         if self.action == "retrieve":
-            try:
-                course_code = self.request.query_params["course"]
-            except KeyError as exc:
-                raise DRFValidationError(
-                    {
-                        "course": _(
-                            "You must specify a course code to get product details."
-                        )
-                    }
-                ) from exc
-
-            queryset = queryset.filter(
-                course_relations__course__code=course_code,
-                course_relations__organizations__isnull=False,
-            )
+            permission_classes = [drf_permissions.AllowAny]
         else:
-            queryset = queryset.filter(course_relations__isnull=False)
+            return super().get_permissions()
 
-        return queryset.select_related("certificate_definition").distinct()
+        return [permission() for permission in permission_classes]
 
-    def get_serializer_context(self):
+    def get_queryset(self):
         """
-        Provide course code to the `ProductSerializer`
+        Custom queryset to limit access to course product relation objects
+        to users who have access to the related course.
         """
-        context = super().get_serializer_context()
 
-        if course := self.request.query_params.get("course"):
-            context.update({"course_code": course})
+        course_code = self.request.query_params.get("course")
 
-        return context
+        if not self.username and not course_code:
+            raise DRFValidationError(
+                {"course": _("You must specify a course code to get product details.")}
+            )
+
+        query_filters = {
+            "organizations__isnull": False,
+        }
+        if course_code:
+            query_filters["course__code"] = course_code
+        else:
+            query_filters["course__accesses__user__username"] = self.username
+
+        return (
+            models.CourseProductRelation.objects.filter(**query_filters)
```

Comment:
it would be more readable to keep the filter on organizations here:
```diff
- models.CourseProductRelation.objects.filter(**query_filters)
+ models.CourseProductRelation.objects.filter(organizations__isnull=False, **query_filters)
```